src/main/java/com/ibm/darpc/DaRPCEndpoint.java

```diff
@@ -70,60 +71,67 @@ public DaRPCEndpoint(DaRPCEndpointGroup<? extends DaRPCEndpoint<R,T>, R, T> endp
 		this.maxinline = rpcGroup.getMaxInline();
 		this.payloadSize = rpcGroup.getBufferSize();
 		this.rawBufferSize = headerSize + this.payloadSize;
-		this.pipelineLength = rpcGroup.recvQueueSize();
-		this.freePostSend = new ArrayBlockingQueue<SVCPostSend>(pipelineLength);
+		this.sendPipelineLength = rpcGroup.sendQueueSize();
+		this.recvPipelineLength = rpcGroup.recvQueueSize();
+		this.freePostSend = new ArrayBlockingQueue<SVCPostSend>(sendPipelineLength);
 		this.pendingPostSend = new ConcurrentHashMap<Integer, SVCPostSend>();
-		this.recvBufs = new ByteBuffer[pipelineLength];
-		this.sendBufs = new ByteBuffer[pipelineLength];
-		this.recvCall = new SVCPostRecv[pipelineLength];
-		this.sendCall = new SVCPostSend[pipelineLength];
+		this.recvBufs = new ByteBuffer[recvPipelineLength];
+		this.sendBufs = new ByteBuffer[sendPipelineLength];
+		this.recvCall = new SVCPostRecv[recvPipelineLength];
+		this.sendCall = new SVCPostSend[sendPipelineLength];
 		this.ticketCount = new AtomicLong(0);
 		this.messagesSent = new AtomicLong(0);
 		this.messagesReceived = new AtomicLong(0);
-		logger.info("RPC client endpoint, with payload buffer size = " + payloadSize + ", pipeline " + pipelineLength);
+		logger.info("RPC client endpoint, with payload buffer size = " + payloadSize + ", send pipeline "
+					+ sendPipelineLength + ", receive pipeline " + recvPipelineLength);
 	}
 	
 	public void init() throws IOException {
-		int sendBufferOffset = pipelineLength * rawBufferSize;
+		int sendBufferOffset = recvPipelineLength * rawBufferSize;
 
 		/* Main data buffer for sends and receives. Will be split into two regions,
 		 * one for sends and one for receives.
 		 */
-		dataBuffer = ByteBuffer.allocateDirect(pipelineLength * rawBufferSize * 2);
-		/* Only do one memory registration with the IB card. */
-		dataMr = registerMemory(dataBuffer).execute().free().getMr();
+		try {
+			dataBuffer = rpcGroup.getWRBuffer(this, sendPipelineLength * rawBufferSize + recvPipelineLength * rawBufferSize);
```

Comment:
I don't understand this. Why make the mempool so complicated but than only get one big chunk out of it and slice them yourself?

Reply:
Ok, I get that each connection might have different send/recv queue size

Reply:
Right, the idea behind this implementation is that the memory pool registers chunks of memory with the card end every endpoint/every QP allocates a pieces of this already registered memory chunk.

The reason, why it is not instantiated in the group is to make the application (Crail) able to read properties, instantiate and "configure" the mempool and then pass to the DaRPC library.

Reply:
I removed this allocator. Instead we use a simple one with fixed blocksize.